feat: add dynamic footer date recipe (#37)
astrolicious/astro-tips.dev: src/content/docs/recipes/dynamic-footer-date-for-static-websites.mdx

---
title: Dynamic footer date for static websites
astroRange: ^4.0.0
---

import { Tabs, TabItem, Steps } from "@astrojs/starlight/components";

Having a copyright notice with the current year such as `© 2024 - All rights reserved` is a common
need (although it doesn't legally enforce anything).

## Server side date: the issue

```astro title="src/components/Footer.astro" {2} "{year}"
---
const year = new Date().getFullYear()
---

<footer>
    <p>&copy; {year} - All rights reserved</p>
</footer>
```

> It works and displays the current year...of when the website was last built! It will only update next time

Reviewer: ```suggestion
However it will only display the year when the website was last built! It will only update next time
```

you rebuild your site.

:::note
This works perfectly fine in SSR because the data is computed on page load.
:::

## The solution

> The solution relies on client-side JavaScript to get the current year when the page loads.

Reviewer: ```suggestion
The solution is to use client-side JavaScript to get the current year when the page loads.
```

<Steps>

> 1. We prepare a date fallback at build time. Indeed, if the user's browser can't run JavaScript,

Reviewer: ```suggestion
1. We prepare a date fallback at build time. If the user's browser can't run JavaScript,
```

we need to show the last known date.

    ```astro
    ---
    const year = new Date().getFullYear()
    ---
    ```

2. We write the markup. We have a span with an id of `footer-year` so that it can be referenced
from our script. It has a default value of the fallback date of step 1.

    ```astro
    <footer>
      <p>&copy; <span id="footer-year">{year}</span> - All rights reserved</p>
    </footer>
    ```

> 3. We write a script to get the `<span>` element by id and update its content to the current year.

Reviewer: ```suggestion
3. In our script we can get the `<span>` element by it's id and update it's content to the current year.
```

> The code differs depending on if you're using View Transitions and your Astro version.

Reviewer: ```suggestion
The code differs depending on your Astro version, and whether you're using View Transitions.
```
